[LoopVectorize] Enable vectorisation of early exit loops with live-outs

david-arm opened the pull request:
This work feeds part of PR https://github.com/llvm/llvm-project/pull/88385, and adds support for vectorising
loops with uncountable early exits and outside users of loop-defined
variables. When calculating the final value from an uncountable early
exit we need to calculate the vector lane that triggered the exit,
and hence determine the value at the point we exited.

All code for calculating the last value when exiting the loop early
now lives in a new vector.early.exit block, which sits between the
middle.split block and the original exit block. Doing this required
two fixes:

1. The vplan verifier incorrectly assumed that the block containing
a definition always dominates the block of the user. That's not true
if you can arrive at the use block from multiple incoming blocks.
This is possible for early exit loops where both the early exit and
the latch jump to the same block.
2. We were adding the new vector.early.exit to the wrong parent loop.
It needs to have the same parent as the actual early exit block from
the original loop.

I've added a new ExtractFirstActive VPInstruction that extracts the
first active lane of a vector, i.e. the lane of the vector predicate
that triggered the exit.

NOTE: The IR generated for dealing with live-outs from early exit
loops is unoptimised, as opposed to normal loops. This inevitably
leads to poor quality code, but this can be fixed up later.

david-arm commented:
PR #88385 is very out of date now, but it also dealt with loops that require a scalar epilogue. I've not attempted to deal with those in this PR, but I am finishing off a downstream patch to handle these too, which will mean we can start vectorising loops in tests/Transforms/LoopVectorizer/multi_early_exit* as well.

I also intend to put up another patch soon that adds support for versioning early exit loops that contain a single load instruction, requiring the correct alignment to ensure safety.

david-arm commented:
Just for reference here is a link to a WIP patch that adds support for versioning of early exit loops with potentially faulting loads: https://github.com/llvm/llvm-project/pull/120603

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3662,6 +3665,13 @@ class VPRegionBlock : public VPBlockBase {
   /// VPRegionBlock.
   VPBlockBase *Exiting;
 
+  /// Hold the Early Exit block of the SEME region, if one exists.
+  VPBasicBlock *EarlyExit;
+
+  /// If one exists, this keeps track of the vector early mask that triggered
+  /// the early exit.
+  VPValue *VectorEarlyExitCond;
+
   /// An indicator whether this region is to generate multiple replicated
```
IIUC those are only needed because currently the final values are computed outside VPlan, right?

Could we instead either force a wide IV and extract the desired lane and add the required recipes directly on VPlan-construction?

Or compute the end value for the user directly in VPlan (see #112145 which computes the end values for the resume phis  #112147 which removes `fixupIVUsers`)?

david-arm replied:
I've rewritten this so that we don't need to stash the `EarlyExit` and `VectorEarlyExitCond` variables. Instead, I've added a `getEarlyExit` that walks through the VPBasicBlocks to find the vector.early.exit block. I also added code in `addUsersInExitBlocks` to work backwards from the early exit block to find the early exit mask. It's uglier, but it does avoid having to track these variables and maintain them during cloning, etc. Hope that's prefereable for you!

fhahn replied:
Thanks, I still need to take a look at the latest update. By forcing a wide IV I meant is ensuring that there always will be a vector induction, from which we should be able to extract the lane for the early exit I think.

One way to ensure that would be to not mark induction updates as uniform if there's an user of the induction outside the loop for early exit loops (something like https://github.com/llvm/llvm-project/commit/5c0d7aad84d52a6afcbd3feb4eeffde5ff94a89d)

fhahn replied:
Just double checking, looking at the latest version it is effectively doing an extract of the last active lane, right?

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -212,7 +212,11 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
         }
 
-        if (!VPDT.dominates(VPBB, UI->getParent())) {
+        // Now that we support vectorising loops with uncountable early exits
+        // we can end up in situations where VPBB does not dominate the exit
+        // block. Only do the check if the user is not in a VPIRBasicBlock.
+        if (!isa<VPIRBasicBlock>(UI->getParent()) &&
+            !VPDT.dominates(VPBB, UI->getParent())) {
```
This isn't a use before def, it's an invalid assumption that the block containing the use has a single predecessor. Now that we support vectorising loops with early exits that assumption is broken, for example the user could be a PHI node.

fhahn replied:
I see; IIUC whether a block has multiple predecessors or not shouldn't really be the issue. The problem is that for a phi we need to check if the incoming value dominates the incoming block. 

We have the same issue for other phi-like recipes, see TODO above (line 210). I think it should be sufficient to skip VPIRInstructions wrapping a phi instruction there as well?

That way we would preserve def-use checks for other recipes in VPIRBasicBlocks.

I also put up https://github.com/llvm/llvm-project/pull/124838 separately to extend def-use verification to phis, which should enable removing the early continue in all cases

david-arm commented on llvm/test/Transforms/LoopVectorize/early_exit_legality.ll:
```diff
@@ -141,7 +140,6 @@ define i64 @loop_contains_load_after_early_exit(ptr dereferenceable(1024) align(
 ; CHECK-LABEL: LV: Checking a loop in 'loop_contains_load_after_early_exit'
 ; CHECK:       LV: Found an early exit loop with symbolic max backedge taken count: 63
 ; CHECK-NEXT:  LV: We can vectorize this loop!
```
Done!

david-arm commented on llvm/test/Transforms/LoopVectorize/multi_early_exit.ll:
```diff
@@ -14,18 +14,18 @@ define i64 @one_uncountable_two_countable_same_exit_phi_of_consts() {
 ; CHECK-NEXT:    [[INDEX:%.*]] = phi i64 [ [[INDEX_NEXT:%.*]], [[LOOP_INC:%.*]] ], [ 3, [[ENTRY:%.*]] ]
 ; CHECK-NEXT:    [[CMP1:%.*]] = icmp ne i64 [[INDEX]], 64
-; CHECK-NEXT:    br i1 [[CMP1]], label [[SEARCH:%.*]], label [[LOOP_END:%.*]]
+; CHECK-NEXT:    br i1 [[CMP1]], label [[SEARCH:%.*]], label [[LOOP_END1:%.*]]
```
Yep, reverted the changes!

david-arm commented on llvm/test/Transforms/LoopVectorize/multi_early_exit_live_outs.ll:
```diff
@@ -12,7 +12,7 @@ define i64 @one_uncountable_two_countable_same_exit() {
 ; CHECK-NEXT:    call void @init_mem(ptr [[P1]], i64 1024)
 ; CHECK-NEXT:    call void @init_mem(ptr [[P2]], i64 1024)
-; CHECK-NEXT:    br label [[LOOP:%.*]]
+; CHECK-NEXT:    br label [[LOOP1:%.*]]
```
Yep, reverted the changes!

david-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2985,6 +3094,23 @@ void InnerLoopVectorizer::fixVectorizedLoop(VPTransformState &State) {
+  // block predecessor of the exit block. This block needs to have the same
+  // parent loop as the exit block itself.
+  VPRegionBlock *VectorRegion = State.Plan->getVectorLoopRegion();
+  if (BasicBlock *OrigEarlyExitBB = Legal->getUncountableEarlyExitBlock()) {
```
I had to fix a bug in `VPBasicBlock::execute` that incorrectly adds some VPBasicBlocks to the wrong parent loop.

fhahn commented on llvm/docs/Vectorizers.rst:
```diff
@@ -405,9 +405,9 @@ Early Exit Vectorization
-(``middle.split`` below) branches to the early exit block. Otherwise
-``middle.block`` selects between the exit block from the latch or the scalar
-remainder loop.
+(``middle.split`` below) branches to the early exit block via an intermediate
+block (``vector.early.exit`` below). Otherwise ``middle.block`` selects between
+the exit block from the latch or the scalar remainder loop.
```
Might be good to also explain why the intermediate block is needed, i.e. to compute the exit value?

david-arm replied:
Done!

fhahn commented on llvm/docs/vplan-early-exit.dot:
```diff
@@ -19,23 +19,27 @@ compound=true
   ]
   N4 -> N5 [ label=""]
-  N4 -> N6 [ label=""]
+  N4 -> N7 [ label=""]
```
Can't comment on the generated file, it looks like the size in pixels changed, just checking if that was intentional?

david-arm replied:
I didn't specifically try to change the size in pixels - I just ran `dot -Tpng ../llvm/docs/vplan-early-exit.dot`. Is there a recommended command for regenerating this? I couldn't see any documentation about this.

david-arm replied:
Thanks for the suggestion! I've regenerated it with that command and hopefully the image looks bigger now.

fhahn replied:
Oh I didn't realize that there were different defaults on different platforms/version. The larger versions looks good, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -8958,6 +8968,9 @@ static void addScalarResumePhis(VPRecipeBuilder &Builder, VPlan &Plan) {
     // start value provides the value if the loop is bypassed.
     bool IsFOR = isa<VPFirstOrderRecurrencePHIRecipe>(VectorPhiR);
     auto *ResumeFromVectorLoop = VectorPhiR->getBackedgeValue();
+    assert(!Plan.getEarlyExit() &&
+           "Cannot handle reductions or first-order recurrences with "
+           "uncountable early exits");
```
This could be added separately, as NFC?

david-arm replied:
Sure, will do although I'll have to check for early exit differently.

david-arm replied:
Created https://github.com/llvm/llvm-project/pull/122732. Will update this PR soon!

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9093,19 +9112,38 @@ addUsersInExitBlocks(VPlan &Plan,
+      if (PredVPBB != MiddleVPBB) {
+        assert(ExitIRI->getParent()->getNumPredecessors() <= 2);
+
+        // Cache the early exit mask
```
```suggestion
        // Lookup and cache the early exit mask.
```

Caching is done to avoid repeated lookups, i.e. compile time perf opt?

david-arm replied:
Yeah that's right. If we don't store the mask in the VPRegionBlock then we have to calculate it manually and I was hoping to avoid constant recalculation of the same value. I could also just calculate this prior to the loop, but then the work is unnecessary if there aren't any early exits.

david-arm replied:
Done

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -501,7 +501,15 @@ void VPBasicBlock::execute(VPTransformState *State) {
     // Register NewBB in its loop. In innermost loops its the same for all
     // BB's.
-    if (State->CurrentParentLoop)
+    if (this == State->Plan->getEarlyExit()) {
+      // If this is the vector early exit block then it has a single successor,
+      // which is the uncountable early exit block of the original loop. The
+      // parent loop for the exit block may not be the same as the parent loop
+      // of the vectorised loop, so we have to treat this differently.
+      Loop *EEL = State->LI->getLoopFor(State->CFG.UncountableEarlyExitBB);
+      if (EEL)
+        EEL->addBasicBlockToLoop(NewBB, *State->LI);
+    } else if (State->CurrentParentLoop)
```
It might be good to frame this more generally if possible.

IIUC the problem here is that now we introduce a VPBB which may be outside the enclosing loop (and in the same loop as the destination VPIRBB.

This is the case when we have a VPBB that only branches to a single exit block and then it must be in the same loop as this VPIRBB. Can we check for that instead?

david-arm replied:
We can do it that way I think, but I guess that requires walking through the list returned by `getExitBlocks()` every time this function is called. I imagine that's quite a bit of compile time cost, but I suppose I can limit this to when `State->CFG.UncountableEarlyExitB` is non-null if that's ok?

huntergr-arm commented on llvm/include/llvm/Transforms/Vectorize/LoopVectorizationLegality.h:
```diff
@@ -407,6 +407,11 @@ class LoopVectorizationLegality {
 
   /// Returns the destination of an uncountable early exiting block.
   BasicBlock *getUncountableEarlyExitBlock() const {
+    if (!HasUncountableEarlyExit) {
```
Is there a reason to have a separate `HasUncountableEarlyExit` variable instead of just checking whether the list of uncountable exit(ing) blocks isn't empty? It just seems that having a single source of truth is better than repeatedly asserting that they agree.

It may also be better to store pairs of (Exiting,Exit) blocks instead of two separate vectors.

david-arm replied:
Split this off into https://github.com/llvm/llvm-project/pull/123219

david-arm replied:
I've left this code unchanged for now, since it's being dealt with in https://github.com/llvm/llvm-project/pull/123219. If that patch lands first I'll rebase this one.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2775,6 +2775,23 @@ BasicBlock *InnerLoopVectorizer::createVectorizedLoopSkeleton(
 
+static bool isValueIncomingFromBlock(BasicBlock *ExitingBB, Value *V,
+                                     Instruction *UI) {
+  PHINode *PHI = dyn_cast<PHINode>(UI);
```
```suggestion
  PHINode *PHI = cast<PHINode>(UI);
```

No need for a separate assert. Might be better to make the parameter a PHINode * instead to make the interface clearer (that this function only works if UI is a phi).

david-arm replied:
I think the reason why it was a dyn_cast prior to my patch is because cast doesn't give as detailed an error message, which is why I kept the dyn_cast here. Having said that, this code won't last long anyway since https://github.com/llvm/llvm-project/pull/112147 removes fixupIVUsers.

david-arm commented:
Rebase

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanCFG.h:
```diff
@@ -306,6 +306,24 @@ template <> struct GraphTraits<VPlan *> {
 };
 
+inline bool VPlan::isExitBlock(VPBlockBase *VPBB) {
+  if (!isa<VPIRBasicBlock>(VPBB) || VPBB->getNumSuccessors() ||
```
Could we just check if it is one of the exit blocks returned by `getExitBlocks`? Might be slightly slower, but shouldn't matter as it is only called at most once per VPBasicBlock.

Of it that is not possible assert that it only returns true if the block is in `getExitBlocks`, to make sure they don't disagree.

Does this need to be defined in `llvm/lib/Transforms/Vectorize/VPlanCFG.h` or could it be moved to VPlan.cpp?

fhahn replied:
It should be fine to define `isExitBlock` in `VPlan.cpp`, which includes `VPlanCFG.h`?

david-arm replied:
Sorry I don't know what's going on here. Not sure if it's my fault that it's editing your comment, or something weird in github?

david-arm commented:
> Just double checking, looking at the latest version it is effectively doing an extract of the last active lane, right?

Hi @fhahn, no it extracts the first active lane by counting the number of trailing zeroes in the mask that triggered the exit. So it's calculating the element corresponding to the first true value in the mask. If it was extracting the last active lane then it would be VF - number of leading zeroes.

fhahn commented:
> > Just double checking, looking at the latest version it is effectively doing an extract of the last active lane, right?
> 
> Hi @fhahn, no it extracts the first active lane by counting the number of trailing zeroes in the mask that triggered the exit. So it's calculating the element corresponding to the first true value in the mask. If it was extracting the last active lane then it would be VF - number of leading zeroes.

Yeah I meant just that it only extracts from last valid lane of the vector IV, not how exactly that is computed (vs computing the end value based on the IV end value)

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9067,15 +9067,21 @@ collectUsersInExitBlocks(Loop *OrigLoop, VPRecipeBuilder &Builder,
   auto *MiddleVPBB = Plan.getMiddleBlock();
   VPBuilder B(MiddleVPBB, MiddleVPBB->getFirstNonPhi());
-  VPTypeAnalysis TypeInfo(Plan.getCanonicalIV()->getScalarType());
+  VPBuilder MiddleB(MiddleVPBB, MiddleVPBB->getFirstNonPhi());
```
`B` appears to be unused now.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9086,19 +9092,38 @@ addUsersInExitBlocks(VPlan &Plan,
-                                             IntegerType::get(Ctx, 32), 1))});
+      VPValue *Ext;
+      VPBasicBlock *PredVPBB =
+          cast<VPBasicBlock>(ExitIRI->getParent()->getPredecessors()[Idx]);
```
I'm not keen on having the predecessors being looked up like this -- I know the operands to the exit phi recipe are added by walking the blocks in `collectUsersInExitBlocks`, but it feels fragile. I won't hold it up for that though, maybe this is something to improve in the future when we add support for additional uncounted exits.

david-arm replied:
I also share your concerns, but at a previous vectoriser call in November @fhahn explained this was quite normal in VPlan that assumptions are made regarding the ordering of successors and predecessors. We already do this when setting up the middle block, etc. It was definitely a problem until I fixed up some code that replaced the VPBasicBlock middle block with a VPIRBasicBlock and maintained the ordering. As we support more exits it may become harder to maintain the ordering during VPlan optimisations.

fhahn replied:
We already have all the required information available in `handleUncountableEarlyExit`. Could we simply update the early exit live-outs directly there and avoid the complication here?

Conceptually it should be fine to have `collectUsersInExitBlocks` only collect exit values with incoming values defined in a loop region; this would allow `handleUncountableEarlyExit` to create the required extract directly for the early exit I think.

I put up https://github.com/llvm/llvm-project/pull/123819 to also avoid collecting live-ins in `collectUsersInExitBlocks` and added an assert that other values to fix are defined in the vector loop region. The assert could then be changed to an early continue.

david-arm replied:
I'm happy to take a look, but as mentioned on https://github.com/llvm/llvm-project/pull/123819 I think I'd like some time to test this out before #123819 lands, in case for some reason it's not possible to do this in `handleUncountableEarlyExit`.

david-arm replied:
I am specifically worried about this code, which I always seem to trip over during a rebase:

```
void VPIRInstruction::execute(VPTransformState &State) {
  assert((isa<PHINode>(&I) || getNumOperands() == 0) &&
         "Only PHINodes can have extra operands");
  for (const auto &[Idx, Op] : enumerate(operands())) {
```

which specifically assumes:

1. The number of operands exactly matches the number of predecessors, and
2. The operands are added in exactly the same order as the order of predecessors.

Anything in the code that breaks these assumptions also breaks early exit loops. By splitting out the code to deal with live-outs in handleUncountableEarlyExit I am worried that we have break 2.

fhahn replied:
> which specifically assumes:
> 
> The number of operands exactly matches the number of predecessors, and
> The operands are added in exactly the same order as the order of predecessors.
> Anything in the code that breaks these assumptions also breaks early exit loops. By splitting out the code to deal > with live-outs in handleUncountableEarlyExit I am worried that we have break 2.

The only case where we could have multiple predecessors at the moment are early-exit loops, so adjusting the exit users where we introduce the new edges should make it more robust I think?

Now that induction exit values can also be handled via extracts, it should be relatively easy to do the update directly in `handleUncountableEarlyExit`. I tried this for the current code (https://github.com/llvm/llvm-project/pull/123819/commits/3d30a341ff0f235181ac6f8d47d83a9f239d3c6f) and it seems to work out relatively nicely at the cost of some minor duplication. But it allows removing the handling of multiple predecessors from `collectUsersInExitBlocks` and co.

david-arm replied:
I have implemented your suggestion for handling the live-outs in handleUncountableEarlyExit. I did this in such a way that the LLVM test suite continues to pass when forcing early exit vectorisation and ignoring potentially faulting loads, and also made sure all the tests in llvm/test/Transforms/LoopVectorize pass. There are still ordering issues that need handling delicately, which is why I had to collect the normal exit values prior to calling handleUncountableEarlyExit, otherwise when verifying the IR we crash due to incoming PHI values being matched up with the wrong incoming block.

fhahn replied:
Ok sounds good, so should https://github.com/llvm/llvm-project/pull/123819 land in preparation?

david-arm replied:
To be honest I didn't base my latest changes off https://github.com/llvm/llvm-project/pull/123819. Ideally I'd prefer to get it all working in this PR so I can be absolutely sure that there are no test failures. At first glance it looks like https://github.com/llvm/llvm-project/pull/123819 will make things more difficult (and code more complicated) for the general case that I care about here. If you take a look at my version in this patch I split out handling of live-outs according to the incoming block which works for all cases (same or different exit blocks), whereas #123819 is trying to handle both the incoming values from the latch and the early exit in handleUncountableEarlyExit.

david-arm replied:
Is the idea behind the changes to handleUncountableEarlyExit in https://github.com/llvm/llvm-project/pull/123819 to have a temporary mechanism to test the special case and defend the code? If so, that would be fine but I think I will need to rewrite that code for the general case that I care about in this PR.

fhahn replied:
Ah I see, instead of handling everything in `handleUncountableEarlyExit` here it requires running the collection early. 

The smallish additional complexity in  `handleUncountableEarlyExit` in https://github.com/llvm/llvm-project/pull/123819 for early exits has the advantage of not having this ordering requirement, but I don't have a strong preference. 

The main benefit from https://github.com/llvm/llvm-project/pull/123819 IMO would be to separately move where we update the currently supported exit users and test/defend the removal of the code in collectUser../addUsers.. independently before adding support for new exit values.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -9086,19 +9092,38 @@ addUsersInExitBlocks(VPlan &Plan,
+        assert(ExitIRI->getParent()->getNumPredecessors() <= 2);
+
+        // Lookup and cache the early exit mask.
+        if (!EarlyExitMask) {
```
Feels like we could have looked up the mask before the loop, but I guess you won't always care about the exit value in an early exit. Some loops may just care that the early exit was taken.

huntergr-arm commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -501,8 +501,15 @@ void VPBasicBlock::execute(VPTransformState *State) {
     // BB's.
-    if (State->CurrentParentLoop)
-      State->CurrentParentLoop->addBasicBlockToLoop(NewBB, *State->LI);
+    Loop *ParentLoop = State->CurrentParentLoop;
```
Is it possible to improve tracking the CurrentParentLoop instead of working around it here?

david-arm replied:
It's possible, although you'll still end up with similar looking code in VPlan::execute where we execute the blocks:

```
  for (VPBlockBase *Block : RPOT)
    Block->execute(State);
```

It's worth bearing in mind that in VPRegionBlock::execute we also set the ParentLoop:

```
  if (!isReplicator()) {
    // Create and register the new vector loop.
    Loop *PrevLoop = State->CurrentParentLoop;
    State->CurrentParentLoop = State->LI->AllocateLoop();
```

and VPRegionBlock is included in that list. So depending upon how the blocks are traversed you may need to cache the previous value each time, i.e. something like:

```
  for (VPBlockBase *Block : RPOT) {
    Loop *OldParentLoop = State->ParentLoop;
    if (Block->getSingleSuccessor() && isExitBlock(Block->getSingleSuccessor()))
      State->ParentLoop = ... parent for exit ...
    Block->execute(State);
    State->ParentLoop = OldParentLoop;
  }
```

I'm honestly not sure if that's any better? The only way to do this neatly I think is if you deal with the VP blocks with a single successor exit block in a different loop, i.e.

```
  for (VPBlockBase *Block : RPOT) {
    if (!Block->getSingleSuccessor() || !isExitBlock(Block->getSingleSuccessor()))
      Block->execute(State);
  }
```

```
  for (VPBlockBase *Block : RPOT) {
    if (Block->getSingleSuccessor() && isExitBlock(Block->getSingleSuccessor())) {
      State->ParentLoop = ... exit block parent loop ...
      Block->execute(State);
    }
  }
```

huntergr-arm commented on llvm/lib/Transforms/Vectorize/VPlanCFG.h:
```diff
@@ -306,6 +306,24 @@ template <> struct GraphTraits<VPlan *> {
+    return false;
+
+  // The block must be a successor of the region block.
+  for (auto *OtherVPBB :
```
nit: `return any_of(vp_depth_first_shallow(RegionBlock->getSingleSuccessor()), ...` would be a bit neater.

david-arm replied:
That requires a lambda function, which I'm not sure is any fewer lines of code, right? Perhaps llvm::is_contained would be more compact than the current version

fhahn reviewed (commented):
> > > > Just double checking, looking at the latest version it is effectively doing an extract of the last active lane, right?
> > > 
> > > 
> > > Hi @fhahn, no it extracts the first active lane by counting the number of trailing zeroes in the mask that triggered the exit. So it's calculating the element corresponding to the first true value in the mask. If it was extracting the last active lane then it would be VF - number of leading zeroes.
> > 
> > 
> > Yeah I meant just that it only extracts from last valid lane of the vector IV, not how exactly that is computed (vs computing the end value based on the IV end value)
> 
> Yeah that's true. There are two different ways of thinking about this:
> 
> 1. We're extracting the lane corresponding to the first true value in the exit mask, or
> 2. We're extracting the last valid lane of the loop-defined vector variable, since values in higher lanes correspond to values beyond the loop exit.
> 
> I've named the new VPInstruction in terms of the former because I thought that more closely matches the operation. Although I can see it depends upon your perspective!

OK great! (Was originally a bit confused as it looks like you modified my original comment but I assumed you meant to post this as new response :)

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3938,6 +3941,22 @@ class VPlan {
   const VPRegionBlock *getVectorLoopRegion() const;
 
+  /// Get the vector early exit block
+  VPBasicBlock *getEarlyExit() {
```
It looks like this is isn't needed any longer?

david-arm replied:
Yep, removed thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -950,6 +957,26 @@ void VPlan::prepareToExecute(Value *TripCountV, Value *VectorTripCountV,
+      VPBB == getScalarHeader())
+    return false;
+
+  VPRegionBlock *RegionBlock = getVectorLoopRegion();
+  if (!RegionBlock)
+    return false;
```
Even if there's no region there could be exit blocks of the VPlan I think

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -950,6 +957,26 @@ void VPlan::prepareToExecute(Value *TripCountV, Value *VectorTripCountV,
+
+  // The block must be a successor of the region block.
+  if (llvm::is_contained(
+          vp_depth_first_shallow(RegionBlock->getSingleSuccessor()), VPBB)) {
```
I am not sure I completely follow the logic here. We return true if `VPBB` is any block after the middle block, basically relying on the checks above to filter out all non-VPIRBBs or VPIRBBs with successors.

Simpler to `return isa<VPIRBasicBlock>(VPBB) && VPBB->getNumSuccessors() == 0;`

david-arm replied:
OK, that's fair enough. I wasn't sure if those two checks would always be enough, but it makes sense.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.cpp:
```diff
@@ -950,6 +957,26 @@ void VPlan::prepareToExecute(Value *TripCountV, Value *VectorTripCountV,
+  // The block must be a successor of the region block.
+  if (llvm::is_contained(
+          vp_depth_first_shallow(RegionBlock->getSingleSuccessor()), VPBB)) {
+    assert(llvm::is_contained(getExitBlocks(), VPBB) &&
```
```suggestion
    assert(is_contained(getExitBlocks(), VPBB) &&
```

david-arm commented:
Rebase + deal with conflicts. NOTE: due to a recent commit (b7286dbef9dc1986860d29e390b092599e1d7db5) `same_exit_block_pre_inc_use1_reverse` now starts vectorising too.

david-arm commented:
Hi @fhahn, the Windows build is complaining about `getExitBlocks` that is called from the new `isExitBlock` function. The error is:

```
C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xutility(1378): error C2280: 'llvm::filter_iterator_impl<llvm::mapped_iterator<IteratorT,FuncTy,BlockTy *>,llvm::VPlan::getExitBlocks::<lambda_38c1aefa419e66528297efd947a86d73>,llvm::detail::fwd_or_bidi_tag_impl<false>::type> &llvm::filter_iterator_impl<llvm::mapped_iterator<IteratorT,FuncTy,BlockTy *>,llvm::VPlan::getExitBlocks::<lambda_38c1aefa419e66528297efd947a86d73>,llvm::detail::fwd_or_bidi_tag_impl<false>::type>::operator =(const llvm::filter_iterator_impl<llvm::mapped_iterator<IteratorT,FuncTy,BlockTy *>,llvm::VPlan::getExitBlocks::<lambda_38c1aefa419e66528297efd947a86d73>,llvm::detail::fwd_or_bidi_tag_impl<false>::type> &)': attempting to reference a deleted function
        with
...
```

It's not immediately obvious to me that this is a problem with `isExitBlock` and possibly a problem with `getExitBlocks` or some assumption about the function is used. I'll investigate this a bit further, but I may just have to remove the assert for now.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -697,14 +697,21 @@ Value *VPInstruction::generate(VPTransformState &State) {
+  case VPInstruction::ExtractFirstActive: {
+    Value *Vec = State.get(getOperand(0));
+    Value *Mask = State.get(getOperand(1));
+    Value *Ctz =
+        Builder.CreateCountTrailingZeroElems(Builder.getInt64Ty(), Mask);
+    return Builder.CreateExtractElement(Vec, Ctz);
```
nit: might be helpful for IR readability to use IR value names here

david-arm replied:
Yep, good suggestion!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -2126,7 +2128,12 @@ bool VPlanTransforms::handleUncountableEarlyExit(
       ExitIRI->addOperand(IncomingFromLatch);
       ExitIRI->extractLastLaneOfOperand(MiddleBuilder);
     }
+
```
nit: unrelated whitespace only change?

fhahn reviewed (commented):
Thanks for the updates, a few small comments left

> Hi @fhahn, the Windows build is complaining about getExitBlocks that is called from the new isExitBlock function. The error is:
> 
> C:\BuildTools\VC\Tools\MSVC\14.29.30133\include\xutility(1378): error C2280: 'llvm::filter_iterator_impl<llvm::mapped_iterator<IteratorT,FuncTy,BlockTy *>,llvm::VPlan::getExitBlocks::<lambda_38c1aefa419e66528297efd947a86d73>,llvm::detail::fwd_or_bidi_tag_impl<false>::type> &llvm::filter_iterator_impl<llvm::mapped_iterator<IteratorT,FuncTy,BlockTy *>,llvm::VPlan::getExitBlocks::<lambda_38c1aefa419e66528297efd947a86d73>,llvm::detail::fwd_or_bidi_tag_impl<false>::type>::operator =(const llvm::filter_iterator_impl<llvm::mapped_iterator<IteratorT,FuncTy,BlockTy *>,llvm::VPlan::getExitBlocks::<lambda_38c1aefa419e66528297efd947a86d73>,llvm::detail::fwd_or_bidi_tag_impl<false>::type> &)': attempting to reference a deleted function
>         with
> ...
> It's not immediately obvious to me that this is a problem with isExitBlock and possibly a problem with getExitBlocks or some assumption about the function is used. I'll investigate this a bit further, but I may just have to remove the assert for now.

Sounds good, it looks like it may not be playing nicely with is_contained for some reason. I also don't have a windows machine to check this.

david-arm committed:
[LoopVectorize] Enable vectorisation of early exit loops with live-outs

This work feeds part of PR #88385, and adds support for vectorising
loops with uncountable early exits and outside users of loop-defined
variables. When calculating the final value from an uncountable early
exit we need to calculate the vector lane that triggered the exit,
and hence determine the value at the point we exited.

All code for calculating the last value when exiting the loop early
now lives in a new vector.early.exit block, which sits between the
middle.split block and the original exit block. Doing this required
the following fix:

* The vplan verifier incorrectly assumed that the block containing
a definition always dominates the block of the user. That's not true
if you can arrive at the use block from multiple incoming blocks.
This is possible for early exit loops where both the early exit and
the latch jump to the same block.

I've added a new ExtractFirstActive VPInstruction that extracts the
first active lane of a vector, i.e. the lane of the vector predicate
that triggered the exit.

NOTE: The IR generated for dealing with live-outs from early exit
loops is unoptimised, as opposed to normal loops. This inevitably
leads to poor quality code, but this can be fixed up later.

david-arm committed:
Attempt to fix Windows build issue with getExitBlocks

david-arm committed:
Remove assert in isExitBlock

david-arm committed:
Address review comments + regen CHECK lines

david-arm commented:
> We have the same issue for other phi-like recipes, see TODO above (line 210). I think it should be sufficient to skip VPIRInstructions wrapping a phi instruction there as well?
> 
> That way we would preserve def-use checks for other recipes in VPIRBasicBlocks.

Yep, I've tried to do that in the latest patch. That will then be superseded by https://github.com/llvm/llvm-project/pull/124838 (or whichever comes first).

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -697,14 +697,21 @@ Value *VPInstruction::generate(VPTransformState &State) {
+    Value *Mask = State.get(getOperand(1));
+    Value *Ctz = Builder.CreateCountTrailingZeroElems(
+        Builder.getInt64Ty(), Mask, true, "first.active.lane");
+    return Builder.CreateExtractElement(Vec, Ctz);
```
nit: could also use a name here

fhahn reviewed (approved):
LGTM ,thanks

david-arm committed:
Add name for early exit value
